!!! FEATURE: Reimplement EditPreviewModes for Neos 9

mficzel opened the pull request:
Since nodes have no context in Neos 9 the edit preview mode cannot be decided by asking the node any more.

This change renovates the `UserInterfaceMode` and the `UserInterfaceModeService` to use php 8 value objects now named `RenderingMode` and `RenderingModeService`. The `renderingModeName` is added as argument to the `NodeController::previewAction` that will pass this as an option to the `Neos\Neos\FusionView` if no renderingModeName the RenderingModeService will determine the currently selected mode of the user. 

The `renderingMode` is then added to the fusionGlobals by the `FusionView` and is used for checking in various places in fusion.

Resolves: #4086
Resolves: #4396
Based on new Fusion globals: #4425

**Upgrade instructions**

The uses of the following fusion expressions have to be adjusted as follows:

- `node.context.live` - `!renderingMode.isEdit` *
- `node.context.inBackend` - `renderingMode.isEdit` *
- `node.context.currentRenderingMode.edit` - `renderingMode.isEdit`
- `node.context.currentRenderingMode.preview` - `renderingMode.isPreview`
- `node.context.currentRenderingMode.name` - `renderingMode.name`

\* this condition should cover the usecase most of the time. For a more accurate representation of "inBackend" use `${renderingMode.isEdit || renderingMode.isPreview}`

**Review instructions**

See #4396 for a full explanation about the decision to make the backend check node independent.

This is an newer approach to pr [#4067](https://github.com/neos/neos-development-collection/pull/4067) 

Follow up Todos:
- The change https://github.com/neos/rector/pull/22 adjusts the rector migrations to use the new renderingMode global. This one should be merged after this pr.
- Remove `Neos.Node.inBackend()` helper after the UI stopped using it
- Consider to adjusting the setting keys `editPreviewModes` and `defaultEditPreviewMode` to `renderingModes` and `defaultRenderingMode` in a separate pr

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mficzel committed:
FEATURE: Implement EditPreviewModes for Neos 9 with the new CR

Since nodes have no context in Neos 9 the edit preview mode cannot be decided by asking the node any more.

This change renovates the `UserInterfaceMode` and the `UserInterfaceModeService` to use php 8 value objects.
The UserInterfaceMode object is then added as globalValue `userInterfaceMode` to the `Neos\Neos\FusionView`.

The uses of the following fusion expressions have to be adjusted as follows:

- `node.context.live` - `!renderingMode.isEdit`
- `node.context.currentRenderingMode.edit` - `renderingMode.isEdit`
- `node.context.currentRenderingMode.preview` - `renderingMode.isPreview`
- `node.context.currentRenderingMode.name` - `renderingMode.name`

Resolves: #4086

mficzel commented:
Since this falls back to UserInterfaceService->findUserInterfaceModeNameByCurrentUser it actually allows to switch the mode in the ui.  🎉

Still think that in the long run the uiMode should be passed to the previewAction (parameter is there) and that even editAction and previewAction should be separates. However both are not necessary to let Neos 9 run and can be added later.

mficzel commented:
A matter to discuss is how to name the properties of the edit preview mode and how they should be accessed. The pr is marked as draft to allow us the short but needed discussion

possible options include: 
- `userInterfaceMode.edit`|`preview`|`live`|`name` - currently implemented and my favorite
- `Neos.uiMode.edit`|`preview`|`live`|`name`
- `Neos.uiMode.edit()`|`preview()`|`live()`|`name()`
- `userInterfaceMode.isEditMode`|`isPreviewMode`|isLive`|`name`

We should discuss and decide this ideally once as this requires adjustments in a dozen fusion files.

mhsdesign commented on Neos.Neos/Classes/Fusion/Helper/NodeHelper.php:
```diff
@@ -95,6 +95,11 @@ public function labelForNode(Node $node): NodeLabelToken
         return new NodeLabelToken($node);
     }
 
+    /**
+     * @param Node $node
+     * @return bool
+     * @deprecated Remove before Neos 9.0 !!! Use ${userInterfaceMode.edit || userInterfaceMode.preview} instead
+     */
```
yes we can remove it in the second beta?

mhsdesign commented:
Hi @mficzel thanks for your 3rd implementation ❤️ !!!

About the naming, we did discuss a few options in the last weekly's (when your were on vacation) and i think most arguments are written down in this thread: https://neos-project.slack.com/archives/C3MCBK6S2/p1693988493145879 and in this issue: https://github.com/neos/neos-development-collection/issues/4086#issuecomment-1665522150

I want to suggest that we both have a little call to discuss all arguments on the table and then we pick the best solution?

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/Prototypes/ContentCollection.fusion:
```diff
@@ -27,7 +27,7 @@ prototype(Neos.Neos:ContentCollection) < prototype(Neos.Fusion:Tag) {
     class.@process.collectionClass = ${Array.push(value, 'neos-contentcollection')}
     data-__neos-insertion-anchor = true
-    data-__neos-insertion-anchor.@if.onlyRenderInBackend = ${Neos.Node.inBackend(documentNode) && node.context.currentRenderingMode.edit}
+    data-__neos-insertion-anchor.@if.onlyRenderInBackend = ${userInterfaceMode.isEdit || userInterfaceMode.isPreview}
```
```suggestion
    data-__neos-insertion-anchor.@if.onlyRenderInBackend = ${userInterfaceMode.isEdit}
```

mhsdesign commented on Neos.Neos/Resources/Private/Fusion/RootCase.fusion:
```diff
@@ -44,12 +43,6 @@ root {
     renderPath = '/page'
   }
 
-  rawContent {
-    @position = 'end 10000'
-    condition = ${Neos.Node.inBackend(documentNode) && documentNode.context.currentRenderingMode.edit}
-    renderPath = '/rawContent'
-  }
-
```
useless definition as covered by line 21.

mficzel commented:
Rename:
- userInterfaceMode -> renderingMode (as in fe there is no ui)
- live -> frontend (to avoid confusion with live workspace)

mhsdesign commented on Neos.Neos/Classes/Controller/Frontend/NodeController.php:
```diff
@@ -234,6 +247,8 @@ public function showAction(string $node, bool $showInvisible = false): void
             $this->handleShortcutNode($nodeAddress, $contentRepository);
         }
 
+        $this->view->setOption('renderingMode', 'frontend');
```
```suggestion
        $this->view->setOption('renderingModeName', 'frontend');
```

mficzel replied:
Yep did the same

mhsdesign commented on Neos.Neos/Classes/Domain/Model/RenderingMode.php:
```diff
@@ -0,0 +1,84 @@
+     * @param string $modeName
+     * @param array<string,mixed> $configuration
+     */
+    public static function createByConfiguration(string $modeName, array $configuration): RenderingMode
```
```suggestion
    public static function createFromConfiguration(string $modeName, array $configuration): RenderingMode
```

mhsdesign commented on Neos.Neos/Classes/Domain/Model/RenderingMode.php:
```diff
@@ -0,0 +1,84 @@
+
+namespace Neos\Neos\Domain\Model;
+
+use Neos\Utility\ObjectAccess;
```
remove import ;)

mhsdesign commented on Neos.Neos/Classes/Controller/Frontend/NodeController.php:
```diff
@@ -120,8 +125,14 @@ class NodeController extends ActionController
      * - For those the CSRF token is validated on the sub-request, so it is safe to be skipped here
      */
-    public function previewAction(string $node): void
+    public function previewAction(string $node, string $renderingModeName = null): void
```
FYI, the ui will currently not use this parameter yet (it could via https://github.com/neos/neos-ui/pull/3411)

But importantly this change allows us to open a page in a new tab independent of the session's value.
Although this would only work for this one page, successive links will currently not have that query param added - the linking service still needs to lear this, but this is out of scope of this pr.

mhsdesign committed:
TASK: Introduce constant `RenderingMode::FRONTEND` for system integrated rendering mode

mhsdesign committed:
TASK: RenderingModeService instances runtime cache

mhsdesign committed:
TASK: Remove half-baked solution of session independent preview's

Otherwise, this feature would only work for this one page, successive links will currently not have that query param added - the linking service still needs to learn this, but we delay this for another time.

mhsdesign reviewed (changes requested):
I found some nitpicks, and proposed a solution for those here: https://github.com/neos/neos-development-collection/pull/4521

mhsdesign commented:
The preview via this link `/neos/redirect?node=user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0%3D__452374b3-3580-2af3-71bd-f9932faea84d` doesnt work yet but because its not implemented in the ui: 

<img width="203" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/3070d501-6f73-4f53-ae47-a5aa68aa79a3">


https://github.com/neos/neos-ui/blob/17361011b1adc5367d77dea632f97df9f7f7703c/Classes/Controller/BackendController.php#L209-L224

mhsdesign commented:
Works so far great inside the ui. I can also confirm that changing the mode via user preferences/ session works and all tabs get the new mode (basically this behavior, which we might want to fix at some point https://github.com/neos/neos-development-collection/issues/965).

But the linking seems a bit off in the ui, and when you open the iframe in a new tab (use the preview endpoint `/neos/preview?node=user-admin__123`)

<img width="947" alt="image" src="https://github.com/neos/neos-development-collection/assets/85400359/564e2d34-b82b-4944-bd1c-06ea7333337f">

Isnt this that issue? https://github.com/neos/neos-development-collection/issues/2347 so no new problem ...

mficzel committed:
TASK: Optimize runtime cache for RenderingModes for readability

mficzel committed:
Merge pull request #4521 from neos/90/editPreviewModeSupport_v3_review

90/edit preview mode support v3 review

mhsdesign reviewed (approved):
We discussed some optional follow ups, but this pr is great as it is and can be merged!

mhsdesign commented:
As the solution was discussed more than 5 times in our weekly’s and additionally @mficzel and me spend time discussing it, i would say this is approach is reviewed/approved enough to be merged with one formal GitHub approval.
